### PR TITLE
Use initialization checker for soundness

### LINKED ISSUE #5854: Unsoundness from not-yet-initialized paths — no null-involved
Here's an unsoundness hole I wasn't aware about, and I can't find in the issue tracker — it involves initialization, but *not* null. Instead of using `null`, we can just select an ill-bounded type from a not-yet-initialized path that's going to crash/loop when we get there — so we get a ClassCastException instead of an abort (exception) or loop.

```scala
scala> new {
     |   val a: String = (((1: Any): b.A): Nothing): String
     |   val b: { type A >: Any <: Nothing } = ???
     | }
java.lang.ClassCastException: java.lang.Integer cannot be cast to scala.runtime.Nothing$
	at rs$line$1$$anon$1.<init>(rs$line$1:2)
```

```scala
scala> new {
     |   val a: String = (((1: Any): b.A): Nothing): String
     |   def loop(): Nothing = loop()
     |   val b: { type A >: Any <: Nothing } = loop()
     | }
     |
java.lang.ClassCastException: java.lang.Integer cannot be cast to scala.runtime.Nothing$
	at rs$line$4$$anon$1.<init>(rs$line$4:2)
	at rs$line$4$.<init>(rs$line$4:5)
```

One can verify that commenting out `val a` gives code that involves no nulls.

### Discussion

We've known for a while that we can inhabit types with bad bounds with expressions that don't return a value — `val v { type A >: Any <: Nothing }  = failing/looping`. But we always reasoned that `v.A` would only be available *after* the crash. The above examples show a loophole we missed.

This loophole is outside the taxonomy of "scaling-up" loopholes in https://www.scala-lang.org/blog/2016/02/17/scaling-dot-soundness.html.

And it doesn't appear in any known DOT variant that I'm familiar with, because it requires *both* paths *and* potentially diverging initialization expressions. If this issue is indeed novel to us, I'll ask other DOT researchers.

Above, the type of `b` has obvious bad bounds, but I assume we can give it non-obvious bad bounds (as usual) to get the same error.

I can see two likely sound potential fixes, but they are not yet realistic.

1. checking strong normalization for value members. Likely sound, pretty restrictive. This could "work" if we required programmers to use some escape hatch (`@total`) when we can't check totality ourselves — the main benefit would be that `@total` documents the danger.
2. forbidding paths starting in `this.f` before `f` is initialized, following @liufengyun et al.'s safe initialization work. I'd expect this to be pretty restrictive, but I don't know.
3. We could also declare these errors out-of-scope.

Regarding 2: In fact, I realized this problem by comparing DOT's and @liufengyun's typing rules for object creation: one declares all fields to be instantly initialized and accessible in the typing context when typing other fields (which is valid if initializers are syntactic values), the other only has in context the actually initialized fields.

### PR DESCRIPTION
**Pull request by liufengyun**
Use initialization checker for soundness

Fix #5854, #4042 and part of #11572

```Scala
-- Warning: tests/init/neg/i11572.scala:8:6 ------------------------------------
8 |  val t: Bounded = new Bounded {
  |      ^
  |      Access non-initialized value t. Calling trace:
  |       -> }	[ i11572.scala:11 ]
  |        -> override type T >: t.T <: t.T	[ i11572.scala:10 ]
1 warning found
```

```Scala
-- Warning: tests/init/neg/i5854.scala:3:6 -------------------------------------
3 |  val b: { type A >: Any <: Nothing } = loop()         // error
  |      ^
  |  Access non-initialized value b. Calling trace:
  |   -> val a: String = (((1: Any): b.A): Nothing): String	[ i5854.scala:2 ]
1 warning found
```

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/transform/init/Effects.scala [hunk 1]
```diff
@@ -12,8 +12,8 @@ import core.Contexts._
 import Potentials._
 
 object Effects {
-  type Effects = Set[Effect]
-  val empty: Effects = Set.empty
+  type Effects = Vector[Effect]
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Effects.scala:15]: Is the change from Set to Vector related to checking `TypeDef`s, or is it an orthogonal improvement?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Effects.scala:15]: It's orthogonal. Set is bad for concatenation, and we can afford to have duplicate entries in summaries.

File: compiler/src/dotty/tools/dotc/transform/init/Summarization.scala [hunk 2]
```diff
@@ -213,6 +214,19 @@ object Summarization {
     else summary
   }
 
+  private def effectsOfType(tp: Type, source: Tree)(implicit env: Env): Effects =
```
Comment (natsukagami) [inline compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:217]: Is this function calculating the effects of constructing an instance of type `tp`?

**Review by natsukagami — COMMENTED** [review]
I don't think I fully understand what `effectsOfType` does, but the rest looks OK!

File: compiler/src/dotty/tools/dotc/transform/init/Summarization.scala (hunk 2 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:217]: It's calculating field access effects in types. In Scala we can have types like `p.f.T`, here we summarize the effects for `p.f`.

**Comment by liufengyun** [comment]
@olhotak Could you also have a look at the changes?

File: compiler/src/dotty/tools/dotc/transform/init/Summarization.scala [hunk 3]
```diff
@@ -213,6 +214,19 @@ object Summarization {
...
+    var summary = Summary.empty
+    val traverser = new TypeTraverser {
+      def traverse(tp: Type): Unit = tp match {
+        case TermRef(_: SingletonType, _) =>
```
Comment (olhotak) [inline compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:221]: Why do we restrict to SingletonType here as a prefix of the TermRef?

What could the prefix be other than a SingletonType and why is it OK to skip those cases?

**Comment by olhotak** [comment]
I have another question above about TermRef/SingletonType.

We should check the various pieces of code in the comments of #50, https://github.com/scala/bug/issues/9633, #4042, #4031, #5854. I think you have some of these already but there seem to be a lot of examples there: there are probably more that should be added as test cases.

Otherwise LGTM.

File: compiler/src/dotty/tools/dotc/transform/init/Summarization.scala (hunk 3 above)
Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:221]: In the community build, the project effpi (which use compiler plugin for session protocol checking) generates types like the following:

```
effpi.examples.demo.audit.types.Pay#replyTo
```

The type looks dubious, but the compiler seems to be fine with it. For such a type, our code will simply recur on the prefix `effpi.examples.demo.audit.types.Pay` and look for term selections recursively.

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:221]: I didn't make it clear:  it's a display problem, it's a TermRef where the prefix is a TypeRef. 

If we only have `_: TermRef`, we will need to do more error-handling in `analyze(tp)` to avoid crash: currently it's pretty strict about what types it may get.

Comment (olhotak) [reply compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:221]: Oh, sorry, yes, this makes sense now.

Then I just wonder why the restriction is for the prefix to be a `SingletonType` rather than the prefix being something like `TermRef | NoPrefix`. Can a `SingletonType` be more things than that and are those things that we want?

Comment (liufengyun) [reply compiler/src/dotty/tools/dotc/transform/init/Summarization.scala:221]: Yes, the prefix can also be `ThisType`, which is usually what we care about. If the prefix is `NoPrefix`, the variable is a local variable --- the analysis currently enforces that all local variables are hot.

**Comment by liufengyun** [comment]
> We should check the various pieces of code in the comments of #50, [scala/bug#9633](https://github.com/scala/bug/issues/9633), #4042, #4031, #5854. I think you have some of these already but there seem to be a lot of examples there: there are probably more that should be added as test cases.

I added some more tests. I didn't include tests that use `null` as a value, as that can be easily handled by null-safety.

**Comment by olhotak** [comment]
> > We should check the various pieces of code in the comments of #50, [scala/bug#9633](https://github.com/scala/bug/issues/9633), #4042, #4031, #5854. I think you have some of these already but there seem to be a lot of examples there: there are probably more that should be added as test cases.
> 
> I added some more tests. I didn't include tests that use `null` as a value, as that can be easily handled by null-safety.

Good point. I will ask @noti0na1 to look at which of the above examples are fixed by `-Yexplicit-nulls` and add them as tests (separately from this PR).
